internal/plugin/grpc_provider.go

```diff
@@ -94,6 +100,45 @@ func (p *GRPCProvider) GetProviderSchema() (resp providers.GetProviderSchemaResp
 	resp.ResourceTypes = make(map[string]providers.Schema)
 	resp.DataSources = make(map[string]providers.Schema)
 
+	// UPBOUND CHANGE: look provider schema up in in-memory cache and on disk.
+	//                 On disk we store the protobuf binary.
+	var cacheFileName string
+	cacheDir := os.Getenv("TF_CACHE_DIR")
+	cacheAddr := p.Addr
+	if cacheAddr.IsZero() {
+		if reattachConfig := p.PluginClient.ReattachConfig(); reattachConfig != nil {
+			id := base36Hash(fmt.Sprintf("%v", reattachConfig))
+			cacheAddr = addrs.NewBuiltInProvider(id)
+			cacheFileName = fmt.Sprintf("provider-getproviderschema-%s.bin", id)
+		}
+	} else {
+		cacheFileName = fmt.Sprintf("provider-getproviderschema-%s.bin", strings.ReplaceAll(cacheAddr.String(), "/", "-"))
+	}
+
+	// check in-memory cache
+	if !cacheAddr.IsZero() {
+		if resp, ok := providers.SchemaCache.Get(cacheAddr); ok {
+			logger.Info("Using cached schema", "file", "in-memory")
+			return resp
+		}
+	}
+
+	// check on disk cache
+	var protoResp *proto.GetProviderSchema_Response
+	if cacheDir != "" && cacheFileName != "" {
```

Comment:
Please see the discussion below for caching the response on disk. For completeness:
```suggestion
	if cacheDir != "" && cacheFileName != "" && p.Addr.IsZero() {
```

We may also rely on the fact that the cache file will be missing.